runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/WatchedConfig.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2021-2022 Aklivity Inc.
+ *
+ * Aklivity licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.aklivity.zilla.runtime.engine.internal.registry;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static org.agrona.LangUtil.rethrowUnchecked;
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+
+public class WatchedConfig
+{
+    private final WatchService watchService;
+    private final Set<WatchKey> watchKeys;
+    private final URL configURL;
+    private byte[] configHash;
+
+    public WatchedConfig(
+        URL configURL,
+        WatchService watchService)
+    {
+        this.watchService = watchService;
+        this.watchKeys = new HashSet<>();
+        this.configURL = configURL;
+    }
+
+    public Map<WatchKey, WatchedConfig> registerPaths()
+    {
+        Map<WatchKey, WatchedConfig> watchedConfigsByKey = new IdentityHashMap<>();
+        Path configPath = Paths.get(configURL.getPath()).toAbsolutePath();
+        try
+        {
+            Set<Path> watchedPaths = new HashSet<>();
+
+            Deque<Path> observablePaths = new LinkedList<>();
+            observablePaths.addLast(configPath);
+
+            if (Files.isSymbolicLink(configPath))
+            {
+                Path targetPath = Files.readSymbolicLink(configPath);
+                if (!targetPath.isAbsolute())
+                {
+                    targetPath = configPath.getParent().resolve(targetPath);
+                }
+                // This is needed so if the symlink contains a symlink chain, those are watched as well
+                observablePaths.addLast(targetPath);
+                // We need to watch for the actual file changes as well
+                watchedPaths.add(toRealPath(targetPath));
```

Comment:
Why do we need to add this here?
Isn't it covered by `if (watchedPaths.add(observablePath))` at line `81` below?
What breaks if we don't use `toRealPath` ?

Reply:
Suppose we have this, as in the test: 
set config: zilla.reconfigure.modify.complex.chain.json -> data/configs/zilla.reconfigure.modify.complex.chain.json
data -> symlink
configs -> realconfigs
Real config is at: symlink/realconfigs/zilla.reconfigure.modify.complex.chain.json
If we remove this line, we don't watch for the file actual config file change, only the symlinks. In our case: 
......../symlink/realconfigs/zilla.reconfigure.modify.complex.chain.json will not be added to watchedPaths

Reply:
It seems this is a nuance to make sure we register on the right path, for all watched paths, so it seems odd to handle it as a special case here, but not on other calls to `watchedPaths.add(...)`.

Perhaps it would be more consistent to call `toRealPath` just before registering with the watch service?

Reply:
That's not good I think, since then we're not watching the symlink's parent itself, rather we'll watch the real path's parent.

Reply:
Yep, this is at the core of my question.

We have different treatment for the actual `configPath` file versus each intermediate step of observable paths. Only the `configPath` is being processed with `toRealPath`.

Does that leave some edge cases open where we are not properly watching the real path of the other observable paths? Perhaps directory-based symlinks are implicitly followed during `Path` registration, making the call `toRealPath` unnecessary for directories?

It seems like the logic in `if (Files.isSymbolicLink(configPath))` block should be applied inside the `while (!observablePaths.isEmpty())` loop to handle some edge cases, like perhaps `config/zilla.yaml => ../current/zilla.yaml`, `current/zilla.yaml => v1/zilla.yaml`, where there are different intermediate real paths of files to watch in different parent directories, not just one for the real path of `config/zilla.yaml`?

Btw, I noticed the `toRealPath` call is currently commented out in the PR.

Reply:
I'm not sure if I got this. One configPath can only have 1 realPath. So in the given example, the realpath of config/zilla.yaml will be (if symlinks created in the root folder): /root/current/v1/zilla.yaml. 
By intermediate real path, what do you mean?

Reply:
In the example of the previous comment, the real path of config/zilla.yaml would resolve through both symlinks, to give v1/zilla.yaml, right?

In that case we would not be watching the intermediate symlink parent for changes, v1 directory, it seems - please double check.

Reply:
In this example, this will be the content of the watchedPaths:

- /Users/barnabas/test/config/zilla.yaml, 
- /Users/barnabas/test/config/../current/zilla.yaml, 
- /Users/barnabas/test/current/v1/zilla.yaml
This means, that we'll watch both symlink's parent dir for changes: 

- for the 1st symlink:  /Users/barnabas/test/config
- for the 2nd symlink: /Users/barnabas/test/config/../current (this will actually be /Users/barnabas/test/current I think, but just to make sure, we cann call .realPath on it)